### ISSUE TITLE
Refactoring HMC/NUTS related functions for better unit testing

### ISSUE THREAD

**Issue by xukai92** [TuringLang/Turing.jl#431]
The current NUTS adaption is buggy (see umbrella issue #324). With careful check on the adaption code by @yebai and me, we believe the bug is probably not come from the adaptation codes themselves, but from the numerical issues inside HMC/NUTS code. To address this, I would like to refactor the codes in HMC/NUTS to do unit testing (HMC samplers was only tested in an integrated setting before).

The main reason stopped us from unit-testing these codes was that the core functions takes `model`, `vi` and `sampler` as inputs, which makes it hard to call them separately. However, as we now actually simply using `vi` as a container inside HMC, i.e. in each HMC step we take variables we need from `vi`, do things, and put it back to `vi`, we can further decompose Turing related variables from core functions of HMC, namely `runmodel`, `sample_momentum`, `find_H` and `find_good_eps`, by passing the corresponding variables or functions fetched from `model`, `vi` and `sampler`. 

Taking `find_H` as an example. The function is currently

```julia
find_H(p::Vector, model::Function, vi::VarInfo, spl::Sampler) = begin
  # NOTE: getlogp(vi) = 0 means the current vals[end] hasn't been used at all.
  #       This can be a result of link/invlink (where expand! is used)
  if getlogp(vi) == 0 vi = runmodel(model, vi, spl) end

  p_orig = p .* spl.info[:wum][:stds]

  H = dot(p_orig, p_orig) / 2 + realpart(-getlogp(vi))
  if isnan(H) H = Inf else H end

  H
end
```

but can be refactoring to something like below

```julia
find_H(p::Vector, logpdf_func_real::Function, stds::Vector) = begin

  if getlogp(vi) == 0 vi = runmodel(model, vi, spl) end

  p_orig = p .* stds
  H = dot(p_orig, p_orig) / 2 + (-logpdf_func_real())
  if isnan(H) H = Inf else H end

  return H

end
```

and calling it by 

```julia
runmodel()
find_H(p, _ -> realpart(Base.invokelatest(model, vi, spl).logp), spl.info[:wum][:stds])
```

in the body of HMC.

In short, the refactoring should be focusing on move the logic of dealing with Turing internal stuff (`vi`, `model` and `sampler`) to the body of HMC, instead of the core functions of HMC. Or in other words one should be able to core functions of HMC completely outside Turing by providing `logpdf_func` or stuff that is needed.

**Comment by yebai** [comment]
@willtebbutt  We discussed something similar a few days ago. Perhaps you can help Kai do design the interface in this refactoring plan.

**Comment by yebai** [comment]
@xukai92 It might be challenging to do this design without refactoring the compiler. Basically, inside your new `find_H` function, we still need to call `runmodel` somewhere to get the joint probability. This means we still need to unpack `p::Vector` into `vi::VarInfo` so that the model can be replayed. We need to somehow refactor the compiler so that it can generate a function that take `p::Vector` as inputs instead of `vi::VarInfo`, which is undesirable as we discussed previous in `VarInfo` related issues.

**Comment by xukai92** [comment]
Emmm, can we just wrap these before calling the core HMC functions? E.g. do you think below would work?

```julia
logpdf_temp(theta) = begin
    vi[spl][:] = theta[:]
    runmodel(model, vi, spl).logp
end

find_H(p, logpdf_temp, spl.info[:wum][:stds])
```

For me, to use the proposed design, it would be similar to what we did to use the ReverseDiff.jl API (https://github.com/yebai/Turing.jl/blob/master/src/core/ad.jl#L116-L119).

**Comment by yebai** [comment]
Yes, this design should work.  Let's go for it!

**Comment by xukai92** [comment]
Closed by https://github.com/TuringLang/Turing.jl/pull/436.
